client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java

```diff
@@ -204,26 +207,29 @@ public List<ShuffleBlockInfo> addRecord(int partitionId, Object key, Object valu
   }
 
   // transform all [partition, records] to [partition, ShuffleBlockInfo] and clear cache
-  public synchronized List<ShuffleBlockInfo> clear() {
+  public List<ShuffleBlockInfo> clear() {
     List<ShuffleBlockInfo> result = Lists.newArrayList();
     long dataSize = 0;
     long memoryUsed = 0;
-    for (Entry<Integer, WriterBuffer> entry : buffers.entrySet()) {
-      WriterBuffer wb = entry.getValue();
-      dataSize += wb.getDataLength();
-      memoryUsed += wb.getMemoryUsed();
-      result.add(createShuffleBlock(entry.getKey(), wb));
-      copyTime += wb.getCopyTime();
+    synchronized (buffers) {
+      final Set<Entry<Integer, WriterBuffer>> entrySet = buffers.entrySet();
+      for (Entry<Integer, WriterBuffer> entry : entrySet) {
+        WriterBuffer wb = entry.getValue();
+        dataSize += wb.getDataLength();
+        memoryUsed += wb.getMemoryUsed();
+        result.add(createShuffleBlock(entry.getKey(), wb));
+        copyTime += wb.getCopyTime();
+      }
+      LOG.info("Flush total buffer for shuffleId[" + shuffleId + "] with allocated["
+          + allocatedBytes + "], dataSize[" + dataSize + "], memoryUsed[" + memoryUsed + "]");
+      buffers.clear();
     }
-    LOG.info("Flush total buffer for shuffleId[" + shuffleId + "] with allocated["
-        + allocatedBytes + "], dataSize[" + dataSize + "], memoryUsed[" + memoryUsed + "]");
-    buffers.clear();
     return result;
   }
 
   // transform records to shuffleBlock
   protected ShuffleBlockInfo createShuffleBlock(int partitionId, WriterBuffer wb) {
-    byte[] data = wb.getData();
+    byte[] data = wb.finalizeAndGetData();
```

Comment:
Why do we need `finalizeAndGetData`?

Reply:
This is a defensive programming. Once the data is gotten, the buffer is aborted.

Reply:
> This is a defensive programming. Once the data is gotten, the buffer is aborted.

Is it necessary? A little weird.

Reply:
I think yes. without this, the other developer may addRecord again after finalize. 

BTW, the similar implementation scatter everywhere

Reply:
Could you paste some example code?

Reply:
https://github.com/apache/flink/blob/master/flink-core/src/main/java/org/apache/flink/util/ResourceGuard.java like this.

Reply:
> https://github.com/apache/flink/blob/master/flink-core/src/main/java/org/apache/flink/util/ResourceGuard.java like this.

It don't have method like `finalizeAndGetData`.

Reply:
Let me remove this part